Implement hasDispatchedInputEvent (event timing)

francovs opened the pull request:
#### c3c59137322cad6287e5fa8b2da50c43db6eed56
<pre>
Implement hasDispatchedInputEvent (event timing)
<a href="https://bugs.webkit.org/show_bug.cgi?id=298663">https://bugs.webkit.org/show_bug.cgi?id=298663</a>
<a href="https://rdar.apple.com/160294430">rdar://160294430</a>

Reviewed by Simon Fraser.

Implements LocalDOMWindow::hasDispatchedInputEvent() and
LocalDOMWindow::setDispatchedInputEvent(), and calls the latter
when queuing the interaction (first-input entry).

This isn&apos;t required for event-timing logic at all, but
hasDispatchedInputEvent() is be required for LCP
(Largest Contentful Paint).

Canonical link: <a href="https://commits.webkit.org/300412@main">https://commits.webkit.org/300412@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/8687b408020fcef0df9d5465be32bd916f8ceed8

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/122427 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/42133 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/32819 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/129023 "Built successfully") | [  ~~🛠 win~~](https://ews-build.webkit.org/#/builders/59/builds/74526 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/124303 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/42852 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/50726 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/93062 "Passed tests") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/59/builds/74526 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/125379 "Passed tests") | [❌ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/34181 "Found 1 new test failure: media/media-video-videorange.html (failure)") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/109624 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/73715 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/33171 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/27781 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/72512 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/103811 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/27992 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/131755 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/49368 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/37574 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/101605 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/49742 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/105845 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/101474 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/25750 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/46850 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/24987 "Passed tests") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/46133 "The change is no longer eligible for processing.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/49226 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/54970 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/48693 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/52043 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/50375 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

smfr commented on Source/WebCore/page/Performance.cpp:
```diff
@@ -321,6 +321,12 @@ void Performance::processEventEntry(const PerformanceEventTimingCandidate& candi
     if (!m_firstInput && candidate.interactionID.value) {
         m_firstInput = PerformanceEventTiming::create(candidate, true);
         queueEntry(*m_firstInput);
+        if (is<Document>(scriptExecutionContext())) {
```
I think there's a better place for this code. Setting `window->setDispatchedInputEvent()` should happen somewhere in the call stack, before we call down into `Performance`, because this state isn't specifically about performance-related stuff. Maybe somewhere in EventDispatcher?

lilyspiniolas commented on Source/WebCore/page/Performance.cpp:
```diff
@@ -321,6 +321,12 @@ void Performance::processEventEntry(const PerformanceEventTimingCandidate& candi
+        if (is<Document>(scriptExecutionContext())) {
+            auto& document = downcast<Document>(*scriptExecutionContext());
+            auto* window = document.window();
+            if (window)
```
```suggestion
            if (RefPtr window = document.window())
```

lilyspiniolas commented on Source/WebCore/page/Performance.cpp:
```diff
@@ -321,6 +321,12 @@ void Performance::processEventEntry(const PerformanceEventTimingCandidate& candi
         m_firstInput = PerformanceEventTiming::create(candidate, true);
         queueEntry(*m_firstInput);
+        if (is<Document>(scriptExecutionContext())) {
+            auto& document = downcast<Document>(*scriptExecutionContext());
```
```suggestion
            RefPtr document = downcast<Document>(*scriptExecutionContext());
```

smfr commented on Source/WebCore/page/LocalDOMWindow.h:
```diff
@@ -283,6 +283,8 @@ class LocalDOMWindow final
 
     using EventTarget::dispatchEvent;
     void dispatchEvent(Event&, EventTarget*);
+    bool hasDispatchedInputEvent() const { return m_hasDispatchedInputEvent; }
```
I think you should add a comment pointing to https://github.com/w3c/event-timing/issues/158, so that readers don't get the wrong understanding about what this is. Also group it with performance API functions, not right next to `dispatchEvent` which reinforces the misunderstanding.

smfr commented on Source/WebCore/page/Performance.cpp:
```diff
@@ -318,9 +318,20 @@ void Performance::processEventEntry(const PerformanceEventTimingCandidate& candi
+    // setDispatchedInputEvent() when an entry with nonzero interactionID has
+    // its duration set, and only if hasDispatchedInputEvent() is false. This, however,
+    // causes inconsistent behavior: a pointerdown event that had its duration assigned
+    // before receiving an interactionID wouldn't qualify for first-input.
```
Is there a GitHub issue you can link to in the comment?

francovs replied:
Added, to https://github.com/w3c/event-timing/issues/159

francovs committed:
Implement hasDispatchedInputEvent (event timing)
https://bugs.webkit.org/show_bug.cgi?id=298663
rdar://160294430

Reviewed by Simon Fraser.

Implements LocalDOMWindow::hasDispatchedInputEvent() and
LocalDOMWindow::setDispatchedInputEvent(), and calls the latter
when queuing the interaction (first-input entry).

This isn't required for event-timing logic at all, but
hasDispatchedInputEvent() is be required for LCP
(Largest Contentful Paint).

Canonical link: https://commits.webkit.org/300412@main
